synapse/push/emailpusher.py

```diff
@@ -234,13 +234,20 @@ def save_last_stream_ordering_and_success(self, last_stream_ordering):
             return
 
         self.last_stream_ordering = last_stream_ordering
-        yield self.store.update_pusher_last_stream_ordering_and_success(
-            self.app_id,
-            self.email,
-            self.user_id,
-            last_stream_ordering,
-            self.clock.time_msec(),
+        pusher_still_exists = (
+            yield self.store.update_pusher_last_stream_ordering_and_success(
+                self.app_id,
+                self.email,
+                self.user_id,
+                last_stream_ordering,
+                self.clock.time_msec(),
+            )
         )
+        if not pusher_still_exists:
+            # The pusher has been deleted while we were processing, so
+            # lets just stop and return.
+            self.on_stop()
```

Comment:
isn't this done when the pusher is deleted?

Reply:
Yeah, I mostly put this in for paranoia to make sure we don't go round the loop again.